engine/src/flutter/lib/web_ui/lib/src/engine/semantics/menus.dart

```diff
@@ -0,0 +1,210 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:ui/ui.dart' as ui;
+
+import 'label_and_value.dart';
+import 'semantics.dart';
+
+/// Indicates a menu element.
+///
+/// Uses aria menu role to convey this semantic information to the element.
+///
+/// Screen-readers takes advantage of "aria-label" to describe the visual.
+class SemanticMenu extends SemanticRole {
+  SemanticMenu(SemanticsObject semanticsObject)
+    : super.withBasics(
+        EngineSemanticsRole.menu,
+        semanticsObject,
+        preferredLabelRepresentation: LabelRepresentation.ariaLabel,
+      ) {
+    setAriaRole('menu');
+  }
+
+  @override
+  void update() {
+    super.update();
+    // Menu items in DropdownButton, PopupMenuButton and MenuAnchor are not the
+    // immediate children of the menu, so we need to set `aria-owns` on menu.
+    // When the menu is open, the tree is still the old one without the menu
+    // item information, so `addOneTimePostUpdateCallback` is called here to get
+    // the latest tree info.
+    semanticsObject.owner.addOneTimePostUpdateCallback(_updateMenuItemId);
+  }
+
+  bool _isMenuItem(SemanticsObject semanticsObject) {
+    return semanticsObject.role == ui.SemanticsRole.menuItem ||
+        semanticsObject.role == ui.SemanticsRole.menuItemCheckbox ||
+        semanticsObject.role == ui.SemanticsRole.menuItemRadio;
+  }
+
+  // Starting from the current semantics node, this method traverses the
+  // semantics tree and collects the menu items by checking whether the role of
+  // the node is [menuItem], then set `aria-owns` attribute to them.
+  void _updateMenuItemId() {
+    final Map<int, SemanticsObject> tree = semanticsObject.owner.semanticsTree;
+    final List<int> ids = [];
+    final int root = semanticsObject.id;
+    final List<int> queue = [];
+    if (tree[root]?.childrenInTraversalOrder != null) {
+      queue.addAll(tree[root]!.childrenInTraversalOrder!);
+    }
+    while (queue.isNotEmpty) {
+      final int child = queue.removeAt(0);
+      if (tree[child] != null && _isMenuItem(tree[child]!)) {
+        ids.add(child);
+      } else {
+        if (tree[child]?.childrenInTraversalOrder != null) {
+          queue.addAll(tree[child]!.childrenInTraversalOrder!);
+        }
+      }
+    }
+
+    String attributeValue = '';
+    for (final int id in ids) {
+      attributeValue = '$attributeValue flt-semantic-node-$id';
+    }
+    setAttribute('aria-owns', attributeValue.trim());
+  }
+
+  @override
+  bool focusAsRouteDefault() => focusable?.focusAsRouteDefault() ?? false;
+}
+
+/// Indicates a menu bar element.
+///
+/// Uses aria menubar role to convey this semantic information to the element.
+///
+/// Screen-readers takes advantage of "aria-label" to describe the visual.
+class SemanticMenuBar extends SemanticRole {
```

Comment:
I have been reading the doc in mdn aria role menubar

```
Menubar interactions should be similar to the typical menu bar interaction in a desktop graphical user interface. In Google Docs, each of those menu items is a menuitem with a popup submenu, so each has an aria-haspopup attribute set to true. The menubar element does not.
```

This seems to me suggesting the menu in menubar should have role menuItem, but I am not sure if this is what it means. I am curious what your interpretation is on the doc.

Reply:
My understanding is menuBar is just a horizontal menu, so they are not a parent-child relationship. And menubar should contain a list of submenu(menuItem with `aria-haspopup` attribute setting to true).